### ISSUE TITLE
Database connection issues

### ISSUE THREAD

**Issue by soulim** [interagent/pliny#230]
Hey guys,

I'm experiencing weird database connection issues. From time to time requests fail because of Sequel exceptions:

* `Sequel::DatabaseDisconnectError: PG::ConnectionBad: PQconsumeInput() SSL connection has been closed unexpectedly`
* `Sequel::DatabaseConnectionError: PG::ConnectionBad: FATAL:  remaining connection slots are reserved for non-replication superuser connections`

I use pliny 0.12 and this app configuration:

* `DB_POOL=10`
* `PUMA_WORKERS=2`
* `PUMA_MIN_THREADS=1`
* `PUMA_MIN_THREADS=10`

The app is running on one Heroku Hobby Dyno and uses Postgres Hobby Dev database. This setup will be improved to production one when the app will be stable.

Locally everything works perfectly fine, but it randomly fails on Heroku. The app could serve 2-3 requests after restart, and then fails, and then could handle requests again. I've tried to figure out what is wrong, but failed and would appreciate any help you could provide.

Thank you!

**Comment by soulim** [comment]
Seems like upgrade to `sequel 4.28.0` solved the problem. Thank you a lot for making great things! :smile:

**Comment by pedro** [comment]
wow, thanks for filling this up – maybe we can bump Sequel in Pliny so no one else runs into this.

How did you realize the bump could help?

**Comment by soulim** [comment]
Can't say what exactly cased the problem :disappointed: But I could provide a bit more details of this story.

Before I've opened this issue on GitHub, I've tried different configurations in order to see if number of pools or possible connections could help, but didn't have any luck.

Also I've tried to reconnect to the database on `before_fork` Puma hook instead of `on_worker_boot`, as it's suggested in [Sequel docs][1]. That didn't help.

Seemed like `select`-requests to database worked perfectly. At least I didn't spot any issue with selects. But the problem appeared regularly when the app tried to insert or update something in the database. It was also weird because according to information from `pg_stat_activity` number of open connections was much less than possible 20.

Before digging deeper into Sequel source code and probably spend noticeable time :wink:, I decided to install the last Sequel version and see how it would work. Somehow that fixed the problem. I'm watching the app for the last few hours, but there's no single connection issue anymore. The app is just running :smile: 

It feels much better when you know exactly where the bug was and how it was fixed. I'm sorry that I can't provide any better information (e.g. link on the closed issue or commit with the fix).

If I could help you with something else, just ping me. I would be glad to help.

[1]: http://sequel.jeremyevans.net/rdoc/files/doc/code_order_rdoc.html#label-Disconnect+If+Using+Forking+Webserver+with+Code+Preloading

**Comment by pedro** [comment]
No worries, this is already super helpful – thanks for sharing!

Out of curiosity, did bumping sequel also update the "pg" gem version?

**Comment by soulim** [comment]
Yes, now the app uses `sequel (4.28.0)` and `pg (0.18.4)`. Do you think the issue might be related to `pg` gem?

**Comment by soulim** [comment]
There's a quick and I hope last update from me. I was wrong thinking that upgrade of `sequel` and `pg` gems helped with the issue. It didn't actually, but there's still a light at the end of the tunnel :smile: 

Yesterday I decided to make sure everything *really* works and did a tiny stress test of the app. I know it does not run on production tier, but the test was not so heavy. So I used `ab` to make a bunch of requests and see how everything goes.

The app failed again with `Sequel::DatabaseDisconnectError: PG::ConnectionBad: PQconsumeInput() SSL connection has been closed unexpectedly` :cry: 

After a couple of hours testing I discovered one thing. The app held strong even under 100 concurrent requests if I ran these requests without long delay. When I did coffee break, got back to work and ran requests again, all of them failed because of this `PG::ConnectionBad` error. Dyno restart always fixed the connection issue, but until I did pause in testing for 10-15 minutes.

This morning I found [this post][1] and [following answer on StackOverflow][2]. Seems like this is exactly my case or at least this explains why I have the issue on Heroku, but never on my local machine.

I didn't know about these rules of hobby tier. The app will be promoted to production tier anyway soon, so I hope the problem will go away too :smile: This is a light I'm looking forward to see :smile: 

Soooo let's keep this issue closed because it's not related to pliny itself. It was just me panicking and searching for help :smile: I still love pliny a lot and thankful for all work you guys did.

[1]: http://geoff.evason.name/2015/01/18/postgres-ssl-connection-has-been-closed-unexpectedly
[2]: http://stackoverflow.com/questions/26792943/heroku-rails-pg-activerecordstatementinvalid-pgconnectionbad-pqconsum/27127603#27127603
